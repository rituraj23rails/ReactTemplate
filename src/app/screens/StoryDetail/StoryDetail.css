.mainContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
    height: 100vh;
}

.headingText {
    font-size: 20px;
    font-weight: bold;
    background-color: #000;
    color: #fff;
    padding: 20px;
}

.formBox {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.textInputField {
    width: 300px;
    height: 50px;
    border: 2px solid #000;
    border-radius: 2px;
    font-size: 20px;
    padding-left: 10px;
    margin-top: 50px;
}

.textInputField:focus {
    outline: none;
}

.submitButton {
    cursor: pointer;
    width: 200px;
    height: 50px;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 2px solid #000;
    margin-top: 50px;
}

.submitButton:hover {
    background-color: green;
    color: #fff;
}

.filterContainer {
    display: flex;
    justify-content: space-between;
    align-items: center;
    cursor: pointer;
    margin-top: 20px;
}

.filterButtons {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 300px;
    height: 50px;
    margin: 0 10px;
    background-color: #000;
}

.buttonText {
    font-size: 20px;
    font-weight: bold;
    color: #fff;
}

.listContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-top: 20px;
    margin-bottom: 100px;
    height: 100vh;
    overflow: scroll;
}

.cellContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    border: 2px solid #000;
    width: 600px;
    padding: 10px 30px;
    margin: 0px auto 20px auto;
}

.dataKey {
    font-size: 15px;
    font-weight: bold;
    margin: 5px 0;
    text-align: left;
}


@media screen and (max-width: 820px) {
    .formBox {
        width: 400px;
        height: 400px;
    }

    .headingText {
        font-size: 18px;
        font-weight: bold;
    }
    
    .textInputField {
        width: 250px;
        height: 50px;
        font-size: 18px;
        padding-left: 10px;
        margin-top: 50px;
    }
    
    .submitButton {
        width: 200px;
        height: 40px;
    }

    .submitButton:hover {
        background-color: green;
        color: #fff;
    }

    .cellContainer {
        width: 500px;
        padding: 10px 30px;
        margin: 0px auto 20px auto;
    }
    
    .dataKey {
        font-size: 12px;
        font-weight: bold;
        margin: 5px 0;
        text-align: left;
    }
    
    .filterButtons {
        display: flex;
        justify-content: center;
        align-items: center;
        width: 200px;
        height: 50px;
        margin: 0 10px;
        background-color: #000;
    }
    
    .buttonText {
        font-size: 15px;
        font-weight: bold;
        color: #fff;
    }

}

@media screen and (max-width: 480px) {
    .formBox {
        width: 300px;
        height: 300px;
    }

    .headingText {
        font-size: 15px;
        font-weight: bold;
    }
    
    .textInputField {
        width: 200px;
        height: 50px;
        font-size: 15px;
        padding-left: 10px;
        margin-top: 50px;
    }
    
    .textInputField:focus {
        outline: none;
    }
    
    .submitButton {
        width: 100px;
        height: 40px;
    }

    .submitButton:hover {
        background-color: green;
        color: #fff;
    }

    .cellContainer {
        width: 300px;
        padding: 10px 30px;
        margin: 0px auto 20px auto;
    }
    
    .dataKey {
        font-size: 12px;
        font-weight: bold;
        margin: 5px 0;
        text-align: left;
        width: 200px;
    }

    .filterContainer {
        display: flex;
        flex-direction: column;
        align-items: center;
        cursor: pointer;
        margin-top: 20px;
    }
    
    .filterButtons {
        display: flex;
        justify-content: center;
        align-items: center;
        width: 200px;
        height: 50px;
        margin: 0 10px;
        background-color: #000;
    }
    
    .buttonText {
        font-size: 15px;
        font-weight: bold;
        color: #fff;
    }
}