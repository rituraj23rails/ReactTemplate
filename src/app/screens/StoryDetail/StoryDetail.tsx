import React, { Component, useRef } from 'react';
import './StoryDetail.css';
import WithRouter from '../../../WithRouter';
import { fetchPostsRequest } from '../../Utils/Fetch';

export interface Props {
    navigate: any;
    id: string;
}

interface S {
    postsData: any;
    originalPostsData: any;
    pageCount: number;
    limit: number;
    loading: boolean;
    pageLoader: boolean;
    pullToRefresh: boolean;
    onEndReachedCalledDuringMomentum: boolean;
    lastLoadCount: number;
    showEmptyView: boolean;
    notFinalLoad: any;
    isFilterCreatedAt: boolean;
    isFilterTitle: boolean;
    intervalId: any
    searchText: string
}

interface SS {
    id: any;
}

class StoryDetail extends Component<Props, S, SS> {
    intervalId: any;
    listInnerRef: any;
    constructor(props: any) {
        super(props);
        this.state = {
            postsData: [],
            originalPostsData: [],
            pageCount: 0,
            limit: 20,
            loading: true,
            pageLoader: false,
            pullToRefresh: false,
            onEndReachedCalledDuringMomentum: true,
            lastLoadCount: 0,
            showEmptyView: false,
            notFinalLoad: null,
            isFilterCreatedAt: false,
            isFilterTitle: false,
            intervalId: null,
            searchText: ''
        };
        this.intervalId = null;
        this.listInnerRef = null;
    };

    componentDidMount = async () => {
        this.fetchStoryData();
        window.addEventListener("scroll", this.onScroll, false);
    }

    componentWillUnmount = () => {
        window.removeEventListener("scroll", this.onScroll, false);
    }

    fetchStoryData = () => {
        this.fetchPostsRequest();
        this.intervalId = setInterval(() => {
            /* istanbul ignore next */
            this.setState({ pageCount: this.state.pageCount + 1 }, () => {
                this.fetchPostsRequest();
            });
        }, 10000);
    }

    //*> Fetch Story Data
    fetchPostsRequest = async () => {
        try {
            let json = await fetchPostsRequest(this.state.pageCount);
            this.setState({
                postsData: this.state.postsData.concat(json.hits),
                originalPostsData: this.state.originalPostsData.concat(json.hits),
                pageLoader: false,
                lastLoadCount: this.state.postsData.concat(json.hits).length,
                onEndReachedCalledDuringMomentum:
                    this.state.postsData.concat(json.hits).length >= this.state.limit
                        ? true
                        : false,
                notFinalLoad:
                    this.state.postsData.concat(json.hits).length >= this.state.limit
                        ? true
                        : false,
            });
        } catch (error) {
        }
    };

    //*> On End Reached of the List Fetch data again
    /* istanbul ignore next */
    onEndReached = () => {
        if (
            this.state.lastLoadCount >= this.state.limit &&
            this.state.notFinalLoad
        ) {
            this.setState(
                {
                    pageLoader: true,
                    pageCount: this.state.pageCount + 1,
                },
                () => {
                    // Then we fetch more data;
                    this.fetchPostsRequest();
                }
            );
        }
    };

    //*> On Navigate to story detail screen
    onPressCellRow = (item: any) => {
        this.props.navigate(`/componentJson`, { state: { rawData: item } });
    };

    filterByCreatedAt = () => {
        clearInterval(this.intervalId);
        this.intervalId = null;
        let localJsonData = this.state.originalPostsData
        const sortByDate = (localJsonData: any) => {
            const sorter = (a: any, b: any) => {
                /* istanbul ignore next */
                return new Date(a.created_at).getTime() - new Date(b.created_at).getTime();
            }
            localJsonData.sort(sorter);
        };
        sortByDate(localJsonData);
        this.setState({ postsData: localJsonData }, () => {
        })
    }

    /* istanbul ignore next */
    filterByTitle = () => {
        clearInterval(this.intervalId);
        this.intervalId = null;
        let localJsonData = this.state.originalPostsData
        localJsonData.sort(function (a: any, b: any) {
            let nameA = a.title.toLowerCase(), nameB = b.title.toLowerCase();
            if (nameA < nameB) { return -1; }
            if (nameA > nameB) { return 1; }
            return 0;
        });
        this.setState({ postsData: localJsonData }, () => {
        })
    }

    clearFilters = () => {
        clearInterval(this.intervalId);
        this.intervalId = null;
        this.setState({ searchText: '', postsData: [], originalPostsData: [] }, () => {
            this.fetchStoryData();
        });
    }
    
    /* istanbul ignore next */
    getSearchData(text: any) {
        clearInterval(this.intervalId);
        this.intervalId = null;
        let localdetail: any = this.state.originalPostsData;
        const newData = localdetail.filter(function (item: any) {
            const itemData = item.title.toLowerCase() + item.url + item.author.toLowerCase();
            const textData = text.toLowerCase()
            return itemData.toString().indexOf(textData) > -1
        });
        if (this.state.searchText.trim().length === 0) {
            this.setState({ postsData: this.state.originalPostsData });
        } else {
            this.setState({ postsData: newData });
        }
    }

    handleChangeTextInput = (event: any) => {
        event?.preventDefault();
        this.setState({ searchText: event.target.value }, () => {
            this.getSearchData(event.target.value)
        });
    }

    onScroll = (e: any) => {
        const bottom = (e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight) || (e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight - 0.5);
        if (bottom) {
            this.onEndReached();
        }
    };

    renderSearch = () => {
        return (
            <div className="formBox">
                <p className="headingText">Story JSON App</p>
                <input data-testid="textInputTestID" type="text" placeholder='Search' className='textInputField' value={this.state.searchText} onChange={(event) => this.handleChangeTextInput(event)} />
            </div>
        )
    }

    renderFilterNSearch = () => {
        return (
            <div className="filterContainer">
                <button data-testid="filterByCreatedAtButton" onClick={() => this.filterByCreatedAt()} className="filterButtons">
                    <p className="buttonText">Filter_by{'\n'}created_at</p>
                </button>
                <button data-testid="clearSearchAndFilterButton" onClick={() => this.clearFilters()} className="filterButtons">
                    <p className="buttonText">Clear Search and Filter</p>
                </button>
                <button data-testid="filterByTitleAtButton" onClick={() => this.filterByTitle()} className="filterButtons">
                    <p className="buttonText">Filter_by{'\n'}title</p>
                </button>
            </div>
        )
    }

    renderPostsListData = () => {
        return (
            <div onScroll={this.onScroll} className="listContainer">
                {this.state.postsData.map((item: any) => {
                    return (
                        <button key={item.title} onClick={() => this.onPressCellRow(item)} className="cellContainer">
                            <p className="dataKey">Title: {item.title}</p>
                            <p className="dataKey">URL: {item.url}</p>
                            <p className="dataKey">Created at: {item.created_at}</p>
                            <p className="dataKey">Author: {item.author}</p>
                        </button>
                    )
                })}
            </div>
        )
    }

    render() {
        return (
            <div className='mainContainer'>
                {this.renderSearch()}
                {this.renderFilterNSearch()}
                {this.renderPostsListData()}
            </div>
        )
    }
};

export default WithRouter(StoryDetail);
export { StoryDetail };
