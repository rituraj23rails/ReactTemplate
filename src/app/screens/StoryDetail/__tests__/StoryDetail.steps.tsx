import { defineFeature, loadFeature } from "jest-cucumber"
import { shallow, ShallowWrapper, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { StoryDetail } from "../StoryDetail";

configure({adapter: new Adapter()});

const screenProps = {
    navigate: jest.fn(),
    id: "StoryDetail"
}

const feature = loadFeature('./src/app/screens/StoryDetail/__tests__/features/StoryDetail-scenarios.feature');

const postsData = [
    {
        title: '',
        url: '',
        created_at: '',
        author: ''
    }
]

defineFeature(feature, (test) => {

    test('User navigates Story Detail Screen', ({ given, when, then }) => {
        let StoryDetailScreenWrapper: ShallowWrapper;
        let StoryDetailScreenInstance: StoryDetail;

        given('I am a User loading Story Detail Screen', () => {
            StoryDetailScreenWrapper = shallow(<StoryDetail {...screenProps} />)
        });

        when('I navigate to the Story Detail Screen', () => {
            StoryDetailScreenInstance = StoryDetailScreenWrapper.instance() as StoryDetail
        });

        then('Story Detail Screen will load with out errors', () => {
            StoryDetailScreenInstance.componentDidMount();
            expect(StoryDetailScreenWrapper).toBeTruthy();
            StoryDetailScreenInstance.setState({ postsData: postsData });
            StoryDetailScreenInstance.onPressCellRow(postsData[0]);
            StoryDetailScreenInstance.onScroll({
                target: {
                    scrollHeight: 1000,
                    scrollTop: 100,
                    clientHeight: 1000,
                }
            })
            expect(StoryDetailScreenWrapper).toMatchSnapshot()
        });

        then('I can enter search input value for stories', () => {
            let inputComponent = StoryDetailScreenWrapper.findWhere((node) => node.prop("data-testid") === "textInputTestID");
            inputComponent.simulate('change', { preventDefault: jest.fn(), target: { value: 'abc' } });
        });
        
        then('I can select the clear filter buttons', () => {
            let buttonComponent = StoryDetailScreenWrapper.findWhere((node) => node.prop("data-testid") === "filterByCreatedAtButton");
            buttonComponent.simulate("click");
            buttonComponent = StoryDetailScreenWrapper.findWhere((node) => node.prop("data-testid") === "clearSearchAndFilterButton");
            buttonComponent.simulate("click");
            buttonComponent = StoryDetailScreenWrapper.findWhere((node) => node.prop("data-testid") === "filterByTitleAtButton");
            buttonComponent.simulate("click");
        });

        then('I can leave the screen with out errors', () => {
            StoryDetailScreenInstance.componentWillUnmount();
            expect(StoryDetailScreenWrapper).toBeTruthy()
            expect(StoryDetailScreenWrapper).toMatchSnapshot()
        });
    });

});
