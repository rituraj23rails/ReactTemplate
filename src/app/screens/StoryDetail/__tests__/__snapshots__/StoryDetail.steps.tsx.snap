// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story Detail Screen User navigates Story Detail Screen 1`] = `
<div
  className="mainContainer"
>
  <div
    className="formBox"
  >
    <p
      className="headingText"
    >
      Story JSON App
    </p>
    <input
      className="textInputField"
      data-testid="textInputTestID"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </div>
  <div
    className="filterContainer"
  >
    <button
      className="filterButtons"
      data-testid="filterByCreatedAtButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Filter_by
        

        created_at
      </p>
    </button>
    <button
      className="filterButtons"
      data-testid="clearSearchAndFilterButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Clear Search and Filter
      </p>
    </button>
    <button
      className="filterButtons"
      data-testid="filterByTitleAtButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Filter_by
        

        title
      </p>
    </button>
  </div>
  <div
    className="listContainer"
    onScroll={[Function]}
  >
    <button
      className="cellContainer"
      key=""
      onClick={[Function]}
    >
      <p
        className="dataKey"
      >
        Title: 
      </p>
      <p
        className="dataKey"
      >
        URL: 
      </p>
      <p
        className="dataKey"
      >
        Created at: 
      </p>
      <p
        className="dataKey"
      >
        Author: 
      </p>
    </button>
  </div>
</div>
`;

exports[`Story Detail Screen User navigates Story Detail Screen 2`] = `
<div
  className="mainContainer"
>
  <div
    className="formBox"
  >
    <p
      className="headingText"
    >
      Story JSON App
    </p>
    <input
      className="textInputField"
      data-testid="textInputTestID"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </div>
  <div
    className="filterContainer"
  >
    <button
      className="filterButtons"
      data-testid="filterByCreatedAtButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Filter_by
        

        created_at
      </p>
    </button>
    <button
      className="filterButtons"
      data-testid="clearSearchAndFilterButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Clear Search and Filter
      </p>
    </button>
    <button
      className="filterButtons"
      data-testid="filterByTitleAtButton"
      onClick={[Function]}
    >
      <p
        className="buttonText"
      >
        Filter_by
        

        title
      </p>
    </button>
  </div>
  <div
    className="listContainer"
    onScroll={[Function]}
  />
</div>
`;
